websocket-codec/src/close.rs

```diff
@@ -0,0 +1,154 @@
+    Restart,
+    /// Indicates that the server is overloaded and the client should either connect
+    /// to a different IP (when multiple targets exist), or reconnect to the same IP
+    /// when a user has performed an action.
+    Again,
+    #[doc(hidden)]
+    Tls,
+    #[doc(hidden)]
+    Reserved(u16),
+    #[doc(hidden)]
+    Iana(u16),
+    #[doc(hidden)]
+    Library(u16),
+    #[doc(hidden)]
+    Bad(u16),
+}
+
+impl CloseCode {
+    /// Check if this `CloseCode` is allowed.
+    #[must_use]
+    pub fn is_allowed(self) -> bool {
+        !matches!(
+            self,
+            Self::Bad(_) | Self::Reserved(_) | Self::Status | Self::Abnormal | Self::Tls
+        )
+    }
+}
+
+impl fmt::Display for CloseCode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let code: u16 = self.into();
+        f.write_str(&code.to_string())
+    }
+}
+
+impl From<CloseCode> for u16 {
+    fn from(code: CloseCode) -> u16 {
+        match code {
+            CloseCode::Normal => 1000,
+            CloseCode::Away => 1001,
+            CloseCode::Protocol => 1002,
+            CloseCode::Unsupported => 1003,
+            CloseCode::Status => 1005,
+            CloseCode::Abnormal => 1006,
+            CloseCode::Invalid => 1007,
+            CloseCode::Policy => 1008,
+            CloseCode::Size => 1009,
+            CloseCode::Extension => 1010,
+            CloseCode::Error => 1011,
+            CloseCode::Restart => 1012,
+            CloseCode::Again => 1013,
+            CloseCode::Tls => 1015,
+            CloseCode::Reserved(code) | CloseCode::Iana(code) | CloseCode::Library(code) | CloseCode::Bad(code) => code,
+        }
+    }
+}
+
+impl<'t> From<&'t CloseCode> for u16 {
+    fn from(code: &'t CloseCode) -> u16 {
+        (*code).into()
+    }
+}
+
+impl From<u16> for CloseCode {
+    fn from(code: u16) -> CloseCode {
+        match code {
+            1000 => CloseCode::Normal,
+            1001 => CloseCode::Away,
+            1002 => CloseCode::Protocol,
+            1003 => CloseCode::Unsupported,
+            1005 => CloseCode::Status,
+            1006 => CloseCode::Abnormal,
+            1007 => CloseCode::Invalid,
+            1008 => CloseCode::Policy,
+            1009 => CloseCode::Size,
+            1010 => CloseCode::Extension,
+            1011 => CloseCode::Error,
+            1012 => CloseCode::Restart,
+            1013 => CloseCode::Again,
+            1015 => CloseCode::Tls,
+            1016..=2999 => CloseCode::Reserved(code),
+            3000..=3999 => CloseCode::Iana(code),
+            4000..=4999 => CloseCode::Library(code),
+            _ => CloseCode::Bad(code),
+        }
+    }
+}
+
+/// A struct representing the close command.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct CloseFrame {
+    /// The reason as a code.
+    pub code: CloseCode,
+    /// The reason as text string.
+    pub reason: String,
```

Comment:
This being a String means it requires allocation, what do you think? tungstenite  uses a `Cow<str>` but I couldn't really get it to work with that

Reply:
Previously I had `String` as such a one-off allocation wasn't a concern in a client application with a long-running connection. It could be a concern for a server with many clients that come and go.

Using `Bytes` here should work, since it will then borrow (via a reference count) from the message's buffer.